tools/depcheck/pkg/cmd/trace/trace.go

```diff
@@ -0,0 +1,301 @@
+		}
+		root = strings.Join([]string{root, slash, "..."}, "")
+		parsedRoots = append(parsedRoots, root)
+	}
+
+	return parsedRoots
+}
+
+func (o *TraceImportsOpts) Validate() error {
+	if len(o.Roots) == 0 {
+		return errors.New("at least one root package must be provided")
+	}
+	if len(o.OutputFormat) > 0 && len(o.Analyze) > 0 {
+		return fmt.Errorf("--output and --analyze are mutually exclusive")
+	}
+	if len(o.OutputFormat) > 0 && o.OutputFormat != "dot" {
+		return fmt.Errorf("invalid output format provided: %s", o.OutputFormat)
+	}
+
+	return nil
+}
+
+// Run execs `go list` on all package entrypoints specified through --root.
+// Each package's ImportPath and non-transitive (immediate) imports are
+// filtered and aggregated. A package is filtered based on whether its ImportPath
+// matches an accepted pattern defined in the set of validPackages.
+// Each aggregated package becomes a node in a generated dependency graph.
+// An edge is created between a package and each of its Imports.
+func (o *TraceImportsOpts) Run() error {
+	args := []string{"list", "--json"}
+	golist := exec.Command("go", append(args, o.Roots...)...)
+
+	r, w := io.Pipe()
+	golist.Stdout = w
+	golist.Stderr = os.Stderr
+
+	pkgs := &PackageList{}
+	go func(list *PackageList) {
+		decoder := json.NewDecoder(r)
+		for {
+			var pkg Package
+			err := decoder.Decode(&pkg)
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error: %v\n", err)
+			}
+
+			list.Add(pkg)
+		}
+	}(pkgs)
+
+	if err := golist.Run(); err != nil {
+		return err
+	}
+
+	g, err := BuildGraph(pkgs, o.excludes)
+	if err != nil {
+		return err
+	}
+
+	if len(o.filters) > 0 {
+		g, err = FilterPackages(g, o.filters)
+		if err != nil {
+			return err
+		}
+	}
+
+	// prune orphans after filtering in order to maximize chance of deleting disconnected
+	// trees from the graph that have been collapsed into their root node by a given filter.
+	g.PruneOrphans()
+
+	if len(o.OutputFormat) > 0 {
+		return o.outputGraph(g)
+	}
+
+	if len(o.Analyze) > 0 {
+		return o.analyzeGraph(g)
+	}
+
+	return nil
+}
+
+func (o *TraceImportsOpts) outputGraph(g graph.Directed) error {
+	if o.OutputFormat != "dot" {
+		return fmt.Errorf("invalid output format provided: %s", o.OutputFormat)
+	}
+
+	data, err := dot.Marshal(g, fmt.Sprintf("%q", strings.Join(o.Roots, ", ")), "", " ", false)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(o.Out, "%v\n", string(data))
+	return nil
+}
+
+// analyzeGraph receives a MutableDirectedGraph and outputs
+// - "Yours": a list of every node in the set unique to the dependency tree of a given target node.
+// - "Mine": a list of every node in the set unique to the dependency tree of the root nodes (set non-overlapping with given target node)
+// - "Ours": a list of every node in the overlapping set between the dependency tree of the root nodes and a given target node
+func (o *TraceImportsOpts) analyzeGraph(g *depgraph.MutableDirectedGraph) error {
```

Comment:
This is really hard to read.  Separate your library methods (things operating on input data), from your command wiring (things combining a set of library methods).  I should have a small method for identifying "which repos are only present because this set of nodes is present".